File: ecos-records/src/test/java/ru/citeck/ecos/records3/test/schema/DtoSchemaTest.java
package ru.citeck.ecos.records3.test.schema;

import lombok.Data;
import org.junit.jupiter.api.Test;
import ru.citeck.ecos.records2.RecordConstants;
import ru.citeck.ecos.records3.RecordsService;
import ru.citeck.ecos.records3.RecordsServiceFactory;
import ru.citeck.ecos.records3.record.atts.schema.annotation.AttName;
import ru.citeck.ecos.records3.record.atts.schema.SchemaAtt;
import ru.citeck.ecos.records3.record.atts.schema.read.DtoSchemaReader;
import ru.citeck.ecos.records3.record.atts.schema.write.AttSchemaWriter;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

public class DtoSchemaTest {

    @Test
    public void test() {

        RecordsServiceFactory serviceFactory = new RecordsServiceFactory();

        DtoSchemaReader dtoSchemaReader = serviceFactory.getDtoSchemaReader();

        List<SchemaAtt> attributesSchema = dtoSchemaReader.read(TestDto.class);
        AttSchemaWriter writer = serviceFactory.getAttSchemaWriter();

        Map<String, String> attributes = writer.writeToMap(attributesSchema);

        assertEquals(4, attributes.size());
        assertEquals("strField", attributes.get("strField"));
        assertEquals("inner{dumb:_null,innerStr}", attributes.get("inner"));

        TestDto dto = new TestDto();
        dto.setInner(new Inner());
        dto.getInner().setInnerStr("innerStr-123");
        //dto.getInner().setAa("bb");
        dto.setNumField(123);
        dto.setStrField("strField-123");

        RecordsService recordsService = serviceFactory.getRecordsService();

        assertEquals(dto, recordsService.getAtts(dto, TestDto.class));
        assertEquals("InnerDisp", serviceFactory.getRecordsService().getAtt(dto, "inner?disp").asText());
        assertEquals("TestDtoDisp", serviceFactory.getRecordsService().getAtt(dto, "?disp").asText());

        assertEquals(dto.attWithCustomName, recordsService.getAtts(dto, CustomNameFieldAtt.class).otherField);

        List<SchemaAtt> constrAtts = serviceFactory.getDtoSchemaReader().read(ConstructorDtoTest.class);
        assertEquals(3, constrAtts.size());
        assertEquals("field0", constrAtts.get(0).getName());
        assertEquals("?disp", constrAtts.get(0).getInner().get(0).getName());
        assertFalse(constrAtts.get(0).getMultiple());
        assertEquals("fieldArr0", constrAtts.get(1).getName());
        assertTrue(constrAtts.get(1).getMultiple());
        assertEquals("?disp", constrAtts.get(1).getInner().get(0).getName());
        assertEquals("intField", constrAtts.get(2).getName());
        assertFalse(constrAtts.get(2).getMultiple());
        assertEquals("?num", constrAtts.get(2).getInner().get(0).getName());
    }

    @Data
    public static class ConstructorDtoTest {

        private final String field0;
        private final List<String> fieldArr0;
        private final int intField;

        public ConstructorDtoTest(@AttName("field0") String field0,
                                  @AttName("fieldArr0") List<String> fieldArr0,
                                  @AttName("intField") int intField) {
            this.field0 = field0;
            this.fieldArr0 = fieldArr0;
            this.intField = intField;
        }
    }

    @Data
    public static class CustomNameFieldAtt {

        @AttName("custom")
        private String otherField;

        private String _null;
    }

    @Data
    public static class TestDto {

        private String strField;
        private int numField;
        private Inner inner;

        @AttName("custom")
        private String attWithCustomName;

        @AttName("?disp")
        public String getDisplayName() {
            return "TestDtoDisp";
        }
    }

    @Data
    public static class Inner {

        private String innerStr;
        @AttName(RecordConstants.ATT_NULL)
        private String dumb;

        @AttName("?disp")
        public String getDisplayName() {
            return "InnerDisp";
        }
    }
}

File: ecos-records/src/test/java/ru/citeck/ecos/records3/test/op/atts/DtoSchemaTest.kt
package ru.citeck.ecos.records3.test.op.atts

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import ru.citeck.ecos.records2.predicate.model.Predicate
import ru.citeck.ecos.records2.predicate.model.VoidPredicate
import ru.citeck.ecos.records3.RecordsServiceFactory
import ru.citeck.ecos.records3.record.atts.schema.annotation.AttName
import ru.citeck.ecos.webapp.api.entity.EntityRef

class DtoSchemaTest {

    @Test
    fun test() {

        val services = RecordsServiceFactory()

        // var atts
        val atts = services.dtoSchemaReader.read(DtoSchema::class.java)

        assertEquals(4, atts.size)

        val attsMap = services.attSchemaWriter.writeToMap(atts)
        assertEquals("?localId", attsMap["id"])
        assertEquals("strField", attsMap["strField"])
        assertEquals("strFieldList[]", attsMap["strFieldList"])
        assertEquals("strFieldSet[]", attsMap["strFieldSet"])

        // val atts
        val atts2 = services.dtoSchemaReader.read(ValDtoSchema::class.java)

        assertEquals(8, atts2.size)

        val attsMap2 = services.attSchemaWriter.writeToMap(atts2)
        assertEquals("?localId", attsMap2["id"])
        assertEquals("strField", attsMap2["strField"])
        assertEquals("strFieldList[]", attsMap2["strFieldList"])
        assertEquals("strFieldSet[]", attsMap2["strFieldSet"])
        assertEquals("enumTest?str", attsMap2["enumTest"])
        assertEquals("enumListTest[]?str", attsMap2["enumListTest"])
        assertEquals("mapTest?json", attsMap2["mapTest"])
        assertEquals("hashMapTest?json", attsMap2["hashMapTest"])

        val valDtoValue = ValDtoSchema(
            "123",
            "456",
            listOf("abc", "def"),
            setOf("ghi", "jkl"),
            ValDtoSchema.EnumClass.SECOND,
            listOf(ValDtoSchema.EnumClass.FIRST, ValDtoSchema.EnumClass.SECOND),
            mapOf(Pair("one", "two")),
            hashMapOf(Pair("one", "two"))
        )
        val valDtoValue2 = services.recordsService.getAtts(valDtoValue, ValDtoSchema::class.java)

        assertEquals(valDtoValue, valDtoValue2)

        val attsMap3 = services.attSchemaWriter.writeToMap(services.dtoSchemaReader.read(DtoSchema2::class.java))
        assertEquals("?id|or('')", attsMap3["id"])

        val valDtoWithDefault = ValDtoWithDefault()
        val valDtoWithDefaultValue = services.recordsService.getAtts(valDtoWithDefault, ValDtoWithDefault::class.java)
        assertEquals(valDtoWithDefault, valDtoWithDefaultValue)

        val withDefaultSchema = services.dtoSchemaReader.read(ValDtoWithDefault::class.java)
        assertEquals(5, withDefaultSchema.size)
    }

    @Test
    fun test2() {

        val services = RecordsServiceFactory()

        val schema = services.dtoSchemaReader.read(TestDtoWithMetaAttAndSet::class.java)
        assertEquals(1, schema.size)
        assertEquals("abc", schema[0].name)
        assertEquals("attributes", schema[0].alias)
        assertTrue(schema[0].multiple)
        assertEquals("?disp", schema[0].inner[0].name)
    }

    class TestDtoWithMetaAttAndSet {
        @AttName("abc[]")
        var attributes: Set<String>? = emptySet()
    }

    data class ValDtoWithDefault(

        val roles: Set<String> = emptySet(),
        val permissions: Set<String> = emptySet(),

        val statuses: Set<String> = emptySet(),
        val condition: Predicate = VoidPredicate.INSTANCE,

        val type: ValDtoSchema.EnumClass = ValDtoSchema.EnumClass.FIRST
    )

    data class ValDtoSchema(
        val id: String?,
        val strField: String?,
        val strFieldList: List<String>?,
        val strFieldSet: Set<String>?,
        val enumTest: EnumClass?,
        val enumListTest: List<EnumClass>?,
        val mapTest: Map<String, Any>?,
        val hashMapTest: HashMap<String, Any>?
    ) {
        enum class EnumClass {
            FIRST,
            SECOND
        }
    }

    class DtoSchema(
        var id: String?,
        var strField: String?,
        var strFieldList: List<String>?,
        var strFieldSet: Set<String>?
    )

    class DtoSchema2(
        var id: EntityRef
    )
}
